pkg/storage/concurrency/concurrency_manager.go

```diff
@@ -0,0 +1,363 @@
+			continue
+		}
+		return g, nil, nil
+	}
+}
+
+// maybeInterceptReq allows the concurrency manager to intercept requests before
+// sequencing and evaluation so that it can immediately act on them. This allows
+// the concurrency manager to route certain concurrency control-related requests
+// into queues and update its optionally internal state based on the requests.
+func (m *managerImpl) maybeInterceptReq(ctx context.Context, req Request) (Response, *Error) {
+	switch {
+	case req.isSingle(roachpb.PushTxn):
+		// If necessary, wait in the txnWaitQueue for the pushee transaction
+		// to expire or to move to a finalized state.
+		t := req.Requests[0].GetPushTxn()
+		resp, err := m.twq.MaybeWaitForPush(ctx, t)
+		if err != nil {
+			return nil, err
+		} else if resp != nil {
+			return makeResponse(resp), nil
+		}
+	case req.isSingle(roachpb.QueryTxn):
+		// If necessary, wait in the txnWaitQueue either for transaction
+		// update or for dependent transactions to change.
+		t := req.Requests[0].GetQueryTxn()
+		return nil, m.twq.MaybeWaitForQuery(ctx, t)
+	default:
+		// TODO(nvanbenschoten): in the future, use this hook to update the
+		// lock table to allow contending transactions to proceed.
+		// for _, arg := range req.Requests {
+		// 	switch t := arg.GetInner().(type) {
+		// 	case *roachpb.ResolveIntentRequest:
+		// 		_ = t
+		// 	case *roachpb.ResolveIntentRangeRequest:
+		// 		_ = t
+		// 	}
+		// }
+	}
+	return nil, nil
+}
+
+// shouldAcquireLatches determines whether the request should acquire latches
+// before proceeding to evaluate. Latches are used to synchronize with other
+// conflicting requests, based on the Spans collected for the request. Most
+// request types will want to acquire latches.
+func shouldAcquireLatches(req Request) bool {
+	switch {
+	case req.ReadConsistency != roachpb.CONSISTENT:
+		// Only acquire latches for consistent operations.
+		return false
+	case req.isSingle(roachpb.RequestLease):
+		// Don't acquire latches for lease requests. These are run on replicas
+		// that do not hold the lease, so acquiring latches wouldn't help
+		// synchronize with other requests.
+		return false
+	}
+	return true
+}
+
+// shouldWaitOnConflicts determines whether the request should wait on locks and
+// wait-queues owned by other transactions before proceeding to evaluate. Most
+// requests will want to wait on conflicting transactions to ensure that they
+// are sufficiently isolated during their evaluation, but some "isolation aware"
+// requests want to proceed to evaluation even in the presence of conflicts
+// because they know how to handle them.
+func shouldWaitOnConflicts(req Request) bool {
+	// TODO what's the best way to define this. There are request types that do
+	// want to wait of locks like (PutRequest and ScanRequest) and then there
+	// are those that don't want to wait on locks like (QueryIntentRequest,
+	// RefreshRequest, and ResolveIntentRequest). Should we define a flag for
+	// this?
```

Comment:
Rather than trying to keep these "internal" request types side-by-side with the "public" KV API (Scan, Put, ...) we should establish a boundary between them. I feel that the flags have not been helping that, and they also tend to obfuscate understanding.

Reply:
I've also grown a little disillusioned about flags for something like this because they're defined so far away from where they're used. I'm curious what you're thinking in terms of a boundary.